src/gt4py/next/program_processors/runners/dace/transformations/redundant_array_removers.py

```diff
@@ -418,3 +419,326 @@ def _get_copy_chain_mode(self, sdfg: dace.SDFG, state: dace.SDFG) -> CopyChainRe
+        #   we should probably lift this restriction.
+        if inner_producer_edge.data.dynamic:
+            return False
+
+        # We only allow arrays that are transients.
+        if not isinstance(temp_desc, dace_data.Array):
+            return False
+        if isinstance(temp_desc, dace_data.View):
+            return False
+        if not temp_desc.transient:
+            return False
+
+        # Check the consumer
+        for consumer_edge in graph.out_edges(temp_node):
+            consumer_sbs: dace_sbs.Range = consumer_edge.data.get_src_subset(consumer_edge, graph)
+            if consumer_sbs is None:
+                return False
+
+            # What is produced must be consumed.
+            if consumer_sbs != produced_sbs:
+                return False
+
+            # Since `temp_node` is eliminated, the consumer must be another AccessNode.
+            # TODO(phimuell): Think of allowing that _some_ consumer are Maps.
+            consumer_node = consumer_edge.dst
+            if not isinstance(consumer_node, dace_nodes.AccessNode):
+                return False
+
+            consumer_desc = consumer_node.desc(sdfg)
+            if not isinstance(consumer_desc, dace_data.Array):
+                return False
+            if isinstance(consumer_desc, dace_data.View):
+                return False
+
+        # `temp_node` will be removed this means it must be single use.
+        if self._single_use_data is None:
+            find_single_use_data = dace_analysis.FindSingleUseData()
+            single_use_data = find_single_use_data.apply_pass(sdfg, None)
+        else:
+            single_use_data = self._single_use_data
+        if temp_node.data not in single_use_data[sdfg]:
+            return False
+
+        return True
+
+    def apply(
+        self,
+        graph: dace.SDFGState,
+        sdfg: dace.SDFG,
+    ) -> None:
+        map_exit: dace_nodes.MapExit = self.map_exit
+        temp_node: dace_nodes.AccessNode = self.temp_node
+        temp_desc = temp_node.desc(sdfg)
+
+        # Understand how the Map writes into the `temp_node`. For this we have to
+        #  look at the subset of the inner edge that writes into the node.
+        assert graph.in_degree(temp_node) == 1
+        outer_producer_edge = next(iter(graph.in_edges(temp_node)))
+        outer_producer_subset = outer_producer_edge.data.get_dst_subset(outer_producer_edge, graph)
+        assert outer_producer_subset is not None
+        assert outer_producer_edge.src_conn.startswith("OUT_")
+        inner_producer_edge = next(
+            iter(
+                inner_producer_edge
+                for inner_producer_edge in graph.in_edges_by_connector(
+                    map_exit, "IN_" + outer_producer_edge.src_conn[4:]
+                )
+            )
+        )
+        inner_producer_subset = inner_producer_edge.data.get_dst_subset(inner_producer_edge, graph)
```

Comment:
We call this logic in both `can_be_applied` and `apply`. Should it be moved to a separate function?